File: backend/src/backend/schemas/downloader.py
from datetime import datetime

from pydantic import BaseModel, Field


class DownloadPreviewRequest(BaseModel):
    url: str = Field(..., description="YouTube or YouTube Music URL")


class DownloadPreviewResponse(BaseModel):
    url: str
    title: str
    artist: str
    album: str | None = None
    duration: float | None = None
    thumbnail_url: str | None = None
    source_bitrate_estimate: int | None = None


class DownloadJobCreate(BaseModel):
    url: str = Field(..., description="YouTube or YouTube Music URL")
    bitrate: int = Field(default=192, description="Target bitrate (128, 192, 256, 320)")
    title_override: str | None = Field(default=None, description="Optional custom title")
    artist_override: str | None = Field(default=None, description="Optional custom artist")
    album_override: str | None = Field(default=None, description="Optional custom album")
    auto_import: bool = Field(
        default=True, description="Automatically import downloaded audio to /music library"
    )


class DownloadJobStatusResponse(BaseModel):
    job_id: str
    url: str
    bitrate: int
    status: str  # pending, downloading, converting, tagging, done, failed
    progress_percent: float = 0.0
    title: str | None = None
    artist: str | None = None
    file_path: str | None = None
    imported_song_id: int | None = None
    error_message: str | None = None
    created_at: datetime
    completed_at: datetime | None = None

File: frontend/src/lib/api/downloader.ts
import { api } from './client';

export interface DownloadPreviewResponse {
	url: string;
	title: string;
	artist: string;
	album?: string | null;
	duration?: number | null;
	thumbnail_url?: string | null;
	source_bitrate_estimate?: number | null;
}

export interface DownloadJobCreate {
	url: string;
	bitrate: number;
	title_override?: string;
	artist_override?: string;
	album_override?: string;
	auto_import: boolean;
}

export interface DownloadJobStatus {
	job_id: string;
	url: string;
	bitrate: number;
	status: 'pending' | 'downloading' | 'converting' | 'tagging' | 'done' | 'failed';
	progress_percent: number;
	title?: string | null;
	artist?: string | null;
	file_path?: string | null;
	imported_song_id?: number | null;
	error_message?: string | null;
	created_at: string;
	completed_at?: string | null;
}

export async function getDownloadPreview(url: string): Promise<DownloadPreviewResponse> {
	return api.post<DownloadPreviewResponse>('/download/preview', { url });
}

export async function startDownloadJob(payload: DownloadJobCreate): Promise<DownloadJobStatus> {
	return api.post<DownloadJobStatus>('/download', payload);
}

export async function getDownloadJobStatus(jobId: string): Promise<DownloadJobStatus> {
	return api.get<DownloadJobStatus>(`/download/jobs/${jobId}`);
}

export async function deleteDownloadJob(jobId: string): Promise<void> {
	return api.delete<void>(`/download/jobs/${jobId}`);
}
